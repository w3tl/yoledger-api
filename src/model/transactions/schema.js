import { accountTypes } from '../accounts';

export default {
  type: 'object',
  required: ['source', 'destination', 'amount', 'createdAt', 'userId'],
  additionalProperties: false,
  properties: {
    _id: {
      bsonType: 'objectId',
    },
    source: {
      type: 'object',
      required: ['name', 'type'],
      additionalProperties: false,
      properties: {
        name: {
          type: 'string',
        },
        type: {
          enum: accountTypes,
        },
      },
    },
    destination: {
      type: 'object',
      required: ['name', 'type'],
      additionalProperties: false,
      properties: {
        name: {
          type: 'string',
        },
        type: {
          enum: accountTypes,
        },
      },
    },
    amount: {
      type: 'number', // TODO: !== 0
    },
    userId: {
      type: 'string',
    },
    createdAt: {
      bsonType: 'date',
    },
  },
};
