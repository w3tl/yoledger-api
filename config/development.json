{
  "credentials": {
    "algorithm": "dev"
  },
  "token": {
    "secret": "jwtsecret",
    "expiresIn": 3000
  }
}
