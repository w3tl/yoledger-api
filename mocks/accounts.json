[
  {
    "name": "Bank card",
    "type": "ASSET",
    "balance": 1500.5
  },
  {
    "name": "Cash",
    "type": "ASSET",
    "balance": 143
  },
  {
    "name": "Train",
    "type": "EXPENSE"
  },
  {
    "name": "Food",
    "type": "EXPENSE"
  }
]
