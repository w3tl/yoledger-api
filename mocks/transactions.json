[
  {
    "source": "Bank card",
    "destination": "Food",
    "amount": 100.5,
    "date": "2018-06-20T00:00:00.000Z"
  },
  {
    "source": "Bank card",
    "destination": "Food",
    "amount": 13,
    "date": "2018-06-21T00:00:00.000Z"
  },
  {
    "source": "Cash",
    "destination": "Train",
    "amount": 124,
    "date": "2018-06-24T00:00:00.000Z"
  }
]
