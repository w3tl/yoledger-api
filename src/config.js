const convict = require('convict');

const config = convict({
  env: {
    doc: 'The application environment.',
    format: ['production', 'development', 'test', 'stage'],
    default: 'development',
    env: 'NODE_ENV',
  },
  port: {
    doc: 'The port to bind.',
    format: 'port',
    default: 3030,
    env: 'GRAPHQL_PORT',
    arg: 'port',
  },
  path: {
    doc: 'Custom path',
    env: 'GRAPHQL_PATH',
    default: 'graphql',
  },
  credentials: {
    algorithm: {
      doc: 'Current crypt algorithm',
      env: 'CRED_ALG',
      default: 'default',
    },
  },
  token: {
    secret: {
      doc: 'Secret used for jwt tokens',
      format: '*',
      default: '',
      sensitive: true,
      env: 'JWT_SECRET',
    },
    expiresIn: {
      doc: 'Expressed in seconds describing a time span',
      format: 'int',
      default: 60, // 1m
      env: 'TOKEN_EXPIRES',
    },
  },
  mongoUri: {
    doc: 'MongoDB connection string',
    format: '*',
    default: 'mongodb://localhost',
    env: 'MONGO_URI',
  },
});

// Load environment dependent configuration
const env = config.get('env');
config.loadFile(`config/${env}.json`);
// Perform validation
config.validate({ allowed: 'strict' });

module.exports = config;
