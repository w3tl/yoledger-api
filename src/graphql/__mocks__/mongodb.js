import {
  users,
  transactions as transactionsJson,
  accounts,
  budgets as budgetsJson,
} from '../mocks';

const mongodb = jest.genMockFromModule('mongodb');

const addCreatedAt = ({ createdAt, ...other }) => ({ ...other, createdAt: new Date(createdAt) });
const addDate = ({ date, ...other }) => ({ ...other, date: new Date(date) });
const transactions = transactionsJson.map(addCreatedAt);
const budgets = budgetsJson.map(addDate);

const mockSession = {
  startTransaction: () => Promise.resolve(),
  commitTransaction: () => Promise.resolve(),
  abortTransaction: () => Promise.reject({ // eslint-disable-line prefer-promise-reject-errors
    errorLabels: ['TransientTransactionError'],
  }),
};

const mockCollections = {
  users: {
    findOne: ({ _id }) => Promise.resolve(users.find(u => u._id === _id)),
    updateOne: () => Promise.resolve({ result: { nModified: 1 } }),
  },
  transactions: {
    find: () => ({
      skip: skip => ({
        limit: limit => ({
          toArray: () => transactions.slice(skip, skip + limit),
        }),
      }),
    }),
    findOne: ({ _id }) => Promise.resolve(transactions.find(t => t._id === _id)),
    insertOne: (transaction) => {
      transactions.push({ _id: '50', ...transaction });
      return Promise.resolve({ insertedId: '50' });
    },
    findOneAndDelete: ({ id }) => Promise.resolve({ value: transactions.find(t => t.id === id) }),
    updateOne: () => Promise.resolve(),
  },
  accounts: {
    find: query => ({
      toArray: () => {
        if ('type' in query) {
          return Promise.resolve(accounts.filter(acc => query.type === acc.type));
        }
        if ('name' in query) {
          if ('$in' in query.name) {
            const result = accounts.filter(({ name }) => query.name.$in.indexOf(name) !== -1);
            return Promise.resolve(result);
          }
        }
        return Promise.resolve([]);
      },
    }),
    findOne: ({ name, _id }) => Promise.resolve(
      accounts.find(acc => name === acc.name || _id === acc._id),
    ),
    insertOne: (account) => {
      accounts.push({ _id: '44', ...account });
      return Promise.resolve({ insertedId: '44' });
    },
    updateOne: () => Promise.resolve(),
  },
  budget: {
    find: ({ $and, date, ...other }) => ({
      toArray: () => {
        if ($and) {
          return Promise.resolve(
            budgets.filter(b => (
              b.account.name === other['account.name']
              && b.date.getTime() >= $and[0].date.$gte.getTime()
              && b.date.getTime() <= $and[1].date.$lte.getTime()
            )),
          );
        }
        return Promise.resolve(budgets.filter(b => b.date.getTime() === new Date(date).getTime()));
      },
    }),
    findOne: ({ account, date }) => {
      if (account) { return Promise.resolve(budgets.find(t => t.account.name === account)); }
      if (date) { return Promise.resolve(budgets.find(t => t.date.getTime() === date.getTime())); }
      return budgets[0];
    },
    updateOne: ({ date, account }, { $set }) => {
      const idx = budgets
        .findIndex(b => b.date.getTime() === date.getTime() && b.account.name === account.name);
      if (idx) {
        budgets[idx] = { ...budgets[idx], ...$set.modificator };
      }
      return Promise.resolve({ upsertedId: budgets[idx]._id });
    },
    aggregate: () => ({
      toArray: () => [
        accounts.find(a => a.name === 'Food'),
        accounts.find(a => a.name === 'Train'),
      ],
    }),
  },
};


mongodb.connection = {
  db: () => ({
    collection: name => mockCollections[name],
  }),
  withSession: callback => callback(mockSession),
};

mongodb.ObjectId = id => String(id);

module.exports = mongodb;
