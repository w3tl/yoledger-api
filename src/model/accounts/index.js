import Model from '../Model';

export const accountTypes = ['ASSET', 'EXPENSE', 'INCOME', 'VENDOR'];

const schema = {
  type: 'object',
  required: ['name', 'type', 'userId'],
  additionalProperties: false,
  properties: {
    _id: {
      bsonType: 'objectId',
    },
    name: {
      type: 'string',
    },
    balance: {
      type: 'number',
    },
    type: {
      enum: accountTypes,
    },
    userId: {
      type: 'string',
    },
  },
};

export default class Account extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.schema = schema;
    this.collName = 'accounts';
  }

  create({
    name, type, balance = 0,
  }) {
    return super.insertOne({
      name,
      type,
      balance,
    });
  }

  addAmount(name, amount, session) {
    return super.updateOne(
      { name, userId: this.userId },
      { $inc: { balance: amount } },
      { session },
    );
  }

  findByName(name, ...props) {
    return super.findOne({ name }, ...props);
  }
}
