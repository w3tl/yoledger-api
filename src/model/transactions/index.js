import Model from '../Model';
import Account from '../accounts';
import schema from './schema';

export default class Transaction extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.collName = 'transactions';
    this.schema = schema;
  }

  async post({
    source, destination, amount, createdAt = new Date(),
  }, session) {
    const Accounts = new Account(this.db, this.userId);
    const sourceAccount = await Accounts.findByName(source);
    const destAccount = await Accounts.findByName(destination);

    if (!sourceAccount) {
      throw new Error(`Account "${source}" not found`);
    }
    if (!destAccount) {
      throw new Error(`Account "${destination}" not found`);
    }

    const transObj = {
      source: { name: sourceAccount.name, type: sourceAccount.type },
      destination: { name: destAccount.name, type: destAccount.type },
      amount,
      createdAt,
    };
    // TODO: apply to budget
    const insertedId = await this.postFunc(transObj, session);
    return insertedId;
  }

  async postFunc(transObj, session) {
    try {
      session.startTransaction();
      const { insertedId } = await super.insertOne(transObj, { session });
      const Accounts = new Account(this.db, this.userId);
      await Accounts.addAmount(transObj.source.name, -transObj.amount, session);
      await Accounts.addAmount(transObj.destination.name, transObj.amount, session);
      await session.commitTransaction();
      return insertedId;
    } catch (error) {
      await session.abortTransaction();
      if (error.errorLabels && error.errorLabels.indexOf('TransientTransactionError') >= 0) {
        // console.log('TransientTransactionError, retrying transaction ...');
        return this.postFunc(transObj, session);
      }
      throw error;
    }
  }

  async unpost(_id, session) {
    try {
      session.startTransaction();
      const { value } = await super.findOneAndDelete(
        { _id },
        { projection: { source: 1, destination: 1, amount: 1 } },
      );
      if (!value) { throw new Error('Transaction not found'); }
      const { source, destination, amount } = value;
      const Accounts = new Account(this.db, this.userId);
      await Accounts.addAmount(source.name, amount, session);
      await Accounts.addAmount(destination.name, -amount, session);
      // TODO: apply to budget
    } catch (e) {
      await session.abortTransaction();
      throw e;
    }

    await session.commitTransaction();
  }

  async update(_id, {
    source, destination, amount, createdAt,
  }, session) {
    const trans = await super.findOne({ _id });
    if (!trans) { throw new Error('Transaction not found'); }

    if (source || destination || amount) {
      await this.unpost(_id, session);
      const transObj = {
        source: source || trans.source.name,
        destination: destination || trans.destination.name,
        amount: amount || trans.amount,
        createdAt: createdAt || trans.createdAt,
      };

      const insertedId = await this.post(transObj, session);
      return insertedId;
    }
    if (createdAt) {
      await super.updateOne({ _id }, { $set: { createdAt } });
      return _id;
    }
    return null;
  }
}
