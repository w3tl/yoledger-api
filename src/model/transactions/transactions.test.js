import { MongoClient, ObjectId } from 'mongodb';
import config from '../../config';
import Transaction from './index';
import Account from '../accounts';
import accounts from '../../../mocks/accounts';
import transactions from '../../../mocks/transactions';

let db;
let connection;
let Trans;
let Accounts;
let transDoc;

const transObj = transactions[0];
const expenseAccount = accounts.find(({ name }) => name === transObj.destination);
const assetAccount = accounts.find(({ name }) => name === transObj.source);

beforeAll(async () => {
  connection = await MongoClient.connect(config.get('mongoUri'), {
    connectWithNoPrimary: true,
    useNewUrlParser: true,
  });
  db = connection.db('dbtransactions');
  Trans = new Transaction(db, 'user1');
  Accounts = new Account(db, 'user1');
  await Accounts.init();
  await Trans.init();
});

afterAll(async () => {
  await connection.close();
});

describe('transaction', () => {
  beforeEach(async () => {
    await Accounts.clear();
    await Trans.clear();
    await Accounts.create(expenseAccount);
    await Accounts.create(assetAccount);
    let _id;
    await connection.withSession(async (session) => {
      _id = await Trans.post(transObj, session);
      transDoc = await Trans.findOne({ _id });
    });
  });

  test('post should create transaction and update the balance of accounts', async () => {
    expect(transDoc).toMatchSnapshot({
      _id: expect.any(ObjectId),
      createdAt: expect.any(Date),
    }, 'Transaction');
    const source = await Accounts.findByName(assetAccount.name);
    expect(source).toMatchSnapshot({
      _id: expect.any(ObjectId),
    }, 'Source account');
    const destination = await Accounts.findByName(expenseAccount.name);
    expect(destination).toMatchSnapshot({
      _id: expect.any(ObjectId),
    }, 'Destination account');
  });

  test('an incorrect account should cause an exception when posting', async () => {
    const transObjToTest = transactions[1];
    expect.assertions(2);
    await expect(connection.withSession(s => Trans.post({ ...transObjToTest, ...{ source: 'bla' } }, s)))
      .rejects
      .toThrowError(/Account .* not found/);

    await expect(connection.withSession(s => Trans.post({ ...transObjToTest, ...{ destination: 'bla' } }, s)))
      .rejects
      .toThrowError(/Account .* not found/);
  });

  test('unpost should cancel transaction', async () => {
    const destinationBefore = await Accounts.findByName(expenseAccount.name);
    const sourceBefore = await Accounts.findByName(assetAccount.name);
    await connection.withSession(session => Trans.unpost(transDoc._id, session));
    const destinationAfter = await Accounts.findByName(expenseAccount.name);
    const sourceAfter = await Accounts.findByName(assetAccount.name);

    expect(sourceAfter.balance).toEqual(sourceBefore.balance + transDoc.amount);
    expect(destinationAfter.balance).toEqual(destinationBefore.balance - transDoc.amount);
  });

  test('when updating accounts, a new transaction must be created', async () => {
    const newSource = { name: 'check', type: 'ASSET', balance: 0 };
    await Accounts.create(newSource);

    let updatedId;
    await connection.withSession(async (session) => {
      updatedId = await Trans.update(transDoc._id, {
        source: newSource.name, amount: 100,
      }, session);
    });
    expect(transDoc._id).not.toEqual(updatedId);
    const destAcc = await Accounts.findByName(expenseAccount.name);
    const sourceAcc = await Accounts.findByName(assetAccount.name);
    const newSourceAcc = await Accounts.findByName(newSource.name);

    expect(sourceAcc.balance).toEqual(assetAccount.balance);
    expect(destAcc.balance).toEqual(100);
    expect(newSourceAcc.balance).toEqual(-100);
  });

  test('an incorrect transaction id should cause an exception when updating or unposting', async () => {
    expect.assertions(2);
    await expect(connection.withSession(s => Trans.unpost('wrongId', s)))
      .rejects
      .toThrowError('Transaction not found');

    await expect(connection.withSession(s => Trans.update('wrongId', {}, s)))
      .rejects
      .toThrowError('Transaction not found');
  });

  test('when updating only createdAt fields, the transaction should only update it', async () => {
    let updatedId;
    await connection.withSession(async (session) => {
      updatedId = await Trans.update(transDoc._id, { createdAt: new Date() }, session);
    });
    expect(transDoc._id).toEqual(updatedId);
  });

  test('nothing to update', async () => {
    await expect(Trans.update(transDoc._id, {}))
      .resolves
      .toBeNull();
  });
});
