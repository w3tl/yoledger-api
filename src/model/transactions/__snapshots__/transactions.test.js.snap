// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction post should create transaction and update the balance of accounts: Destination account 1`] = `
Object {
  "_id": Any<ObjectID>,
  "balance": 100.5,
  "name": "Food",
  "type": "EXPENSE",
  "userId": "user1",
}
`;

exports[`transaction post should create transaction and update the balance of accounts: Source account 1`] = `
Object {
  "_id": Any<ObjectID>,
  "balance": 1400,
  "name": "Bank card",
  "type": "ASSET",
  "userId": "user1",
}
`;

exports[`transaction post should create transaction and update the balance of accounts: Transaction 1`] = `
Object {
  "_id": Any<ObjectID>,
  "amount": 100.5,
  "createdAt": Any<Date>,
  "destination": Object {
    "name": "Food",
    "type": "EXPENSE",
  },
  "source": Object {
    "name": "Bank card",
    "type": "ASSET",
  },
  "userId": "user1",
}
`;
